exporter/awsemfexporter/emf_exporter.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsemfexporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.opentelemetry.io/collector/translator/internaldata"
+	"go.uber.org/zap"
+)
+
+type emfExporter struct {
+	//Each (log group, log stream) keeps a separate Pusher because of each (log group, log stream) requires separate stream token.
+	groupStreamToPusherMap map[string]map[string]Pusher
+	svcStructuredLog       LogClient
+	config                 configmodels.Exporter
+	logger                 *zap.Logger
+
+	pusherMapLock sync.Mutex
+	retryCnt      int
+}
+
+// New func creates an EMF Exporter instance with data push callback func
+func New(
+	config configmodels.Exporter,
+	params component.ExporterCreateParams,
+) (component.MetricsExporter, error) {
+	if config == nil {
+		return nil, errors.New("emf exporter config is nil")
+	}
+
+	logger := params.Logger
+	// create AWS session
+	awsConfig, session, err := GetAWSConfigSession(logger, &Conn{}, config.(*Config))
+	if err != nil {
+		return nil, err
+	}
+
+	// create CWLogs client with aws session config
+	svcStructuredLog := NewCloudWatchLogsClient(logger, awsConfig, session)
+
+	emfExporter := &emfExporter{
+		svcStructuredLog: svcStructuredLog,
+		config:           config,
+		retryCnt:         *awsConfig.MaxRetries,
+		logger:           logger,
+	}
+	emfExporter.groupStreamToPusherMap = map[string]map[string]Pusher{}
+
+	return emfExporter, nil
+}
+
+func (emf *emfExporter) pushMetricsData(_ context.Context, md pdata.Metrics) (droppedTimeSeries int, err error) {
+	expConfig := emf.config.(*Config)
+	logGroup := "/metrics/default"
+	logStream := "otel-stream"
+	// override log group if customer has specified Resource Attributes service.name or service.namespace
+	putLogEvents, totalDroppedMetrics, namespace := generateLogEventFromMetric(md)
+	if namespace != "" {
+		logGroup = fmt.Sprintf("/metrics/%s", namespace)
+	}
+	// override log group if found it in exp configuration, this configuration has top priority. However, in this case, customer won't have correlation experience
```

Comment:
Shouldn't we give resource name / namespace highest priority? And use the config values as the defaults if not found?

Reply:
So our zap config for LogGroupName / LogStreamName could just default to "/metrics/default" and "otel-stream"

Reply:
We want to give customers the ability to override logGroup/logStream by simply modifying the yaml config. If the resource attribute has the highest priority, it's difficult for customers to override it.